### DISCUSSION TITLE
GRADKIENT_POLYLINE issues

### DISCUSSION THREAD

**Discussion by bigdigital** [zepp-health/zeppos-forum#22] [Developer Feedback]
Hi! For my project, i need to display a custom graphs.  I found a very suitable widget GRADKIENT_POLYLINE to draw a graph. But as is can see this widget has a lot of bugs so it is almost impossible to use it (at least on API v1).  The test was made on GTR 3 Pro v1.0.2 simulator 
 and on the real Amazfit GTR 3 Pro v 8.43.5.1
I made a repository with a test that can show the issue.

1. The widget output buffer is stripped:
![1PnVyGi](https://user-images.githubusercontent.com/4833779/207839106-ae3aa156-0fc3-4e28-9196-ba70d107dc5b.png)
This issue even noticeable on GRADKIENT_POLYLINE 
API example https://docs.zepp.com/docs/reference/device-app-api/newAPI/ui/widget/GRADIENT_POLYLINE/  ![4ocxml9](https://user-images.githubusercontent.com/4833779/207838934-344f9f61-c6cf-4e35-be08-999bcbca64ce.png).  
 Link to code  example https://github.com/bigdigital/GRADKIENT_POLYLINE/blob/master/page/gtr3-pro/home/index.page.js

2. If creating two widgets with the same coordinates and size, the second widget will clear the output from the first  
![P9h0eO6](https://user-images.githubusercontent.com/4833779/207839539-4dfd8869-5660-48c3-98df-d1f5c2c795e2.png)
Link to code example https://github.com/bigdigital/GRADKIENT_POLYLINE/blob/Issue2_graph_overlapp/page/gtr3-pro/home/index.page.js

3. If create two widgets, the second widget will replace all properties for the first (color, lineData, etc)
![j3eDIek](https://user-images.githubusercontent.com/4833779/207840414-3c70192c-1839-41f8-b9d2-4a1b1d3d49d0.png)
 Link to code  example https://github.com/bigdigital/GRADKIENT_POLYLINE/blob/issue3_widget_properties_repalacing/page/gtr3-pro/home/index.page.js

4.  Although the addPoint function is not documented, it somehow works. But for me it is expected it should put only dots to the canvas, but instead, it outputs a dots connected with lines 
![FsDutRi](https://user-images.githubusercontent.com/4833779/207840705-5a3f68d7-c0f9-4861-a49b-bbb461238472.png)
 Link to code  example  https://github.com/bigdigital/GRADKIENT_POLYLINE/blob/issue4_dot_display/page/gtr3-pro/home/index.page.js

**Comment by bigdigital** [comment]
Just figured out that issues 2 and 3, most possibly are the same issue. Because of the issue 3 (all line styles become the same for 1 widget), so the line has the same positions

**Reply by cuiyedong** [reply]
Thank you for your feedback. According to the feedback bug information, we have made targeted fixes on the code.

1. bug 1 : polyline is stripped
    reason : polyline canvas's rect is not match screen size,it is smaller than screen size, so ,the polyline draw is stripped.
    fix : adjust canvas rect.
    
![image](https://user-images.githubusercontent.com/52225002/211963411-511c93a6-45bc-4687-a4e2-35e8fa1d4358.png)

2. bug 2/ bug 3 : second polyline properties was replaced
   reason : clone element attributes are not fully isolated
   fix : Create private properties for each element, isolated from each other
   
![image](https://user-images.githubusercontent.com/52225002/211964828-7ba56a41-c0c8-484c-bf86-ef4dbe356c1c.png)

3. bug 4: addPoint 
in this widget,addPoint means that add a new data to the line chart array,not a isolated point。we will detailed in the docs@MyoungXUE

**Reply by twinko** [reply]
Hi @cuiyedong  sorry for asking again but do you have a rough estimate when this change will be released?
Regards

**Comment by MyoungXUE** [comment]
Thank you for your support of the Developer Forum.

I will test the mentioned code cases and confirm with relevant colleagues, please be patient for a while.

**Comment by MyoungXUE** [comment]
Thank you very much for your feedback, this is a bug that needs to be fixed by the firmware!

Your suggestions are very good and we will take them on board, please be patient and wait for our fix!

**Comment by twinko** [comment]
Hey everyone,(@cuiyedong @MyoungXUE @Xuemuyang) thank you for the solution, is there at least a rough estimate, when this fix get pushed to the devices?
